### ISSUE TITLE
Editable menus

### ISSUE THREAD

**Issue by MichaelMackus** [LambdaHack/LambdaHack#226]
I noticed it seems non-intuitive/difficult to edit the menus in any sort of non-trivial way. It would be nice to even have a "customSplashMenu" setting or something that you could set to a multi-line ASCII string for a simple way to do a custom main-menu. I think this was supported in an older version of the engine.

**Comment by MichaelMackus** [comment]
Also, things like custom prompts/submenus are hard for me to grok. I don't see how you could even add a "What is your name?" prompt for example (much less "Race" and "Class" menus).

**Comment by Mikolaj** [comment]
Yes, you are right. This is heavily under-developed (see #132, perhaps what's proposed there would be an improvement?) and complexity of colours and fonts and highlightable and inert parts (see the help screens for what needs to be supported) leaks through any weak abstractions we might have. 

I don't have experience with widget libraries, but perhaps we should just have several ready widgets (but I guess in libraries they are usually much less general that what we need?). Or perhaps Brick can be an inspiration: https://hackage.haskell.org/package/brick. However, even the single most important widget that we have, a menu, is too complex right now, both in inputs and outputs, just as you experienced.

Main menu is now more constrained, because it displays the blurb, which needs to be rendered differently depending on whether square or half-width rectangular font is used. Also, it corresponds visually to the scenario choice screens. Still, configurability and, in particular, a configurable background ASCII art, would be welcome. Perhaps after we simplify menus and define other widgets to avoid the current painful compllexity.

You are right, we don't have any text entering widget. And it's probably not trivial at all to add. The closest thing we have is number entering (one of 3 widgets we have, total?)

https://github.com/LambdaHack/LambdaHack/blob/95547fbb0549ba3e32a36911f60e0834f112dd3f/engine-src/Game/LambdaHack/Client/UI/HandleHelperM.hs#L360

That definitely deserves its own issue. Perhaps it could be tested in the sample LambdaHack game by enabling edits to hero names (with initial defaults already in config). That would also enable #71.

**Comment by Mikolaj** [comment]
However "Race" and "Class" menus are relatively easy to do. E.g. in exactly the way `tweak convenience settings`/`mark suspect terrain` can be currently cycled among `untried`, `all` and `none`. Or `start new game`/`scenario`, which displays whatever you want on the right, depending on which option is chosen. I guess, for now, the best way to implement class would be as an "organ" that is displayed in `@` menu, representing the accumulated training residing in the brain. Perhaps class could even be the main organ, the body of the actor.

**Comment by MichaelMackus** [comment]
In terms of widgets, it looks like SDL has a [tutorial regarding TextInput widgets](https://wiki.libsdl.org/Tutorials/TextInput). However, the library functions are not fully functional and there are no "build-in" widgets currently supported. I did use the TextInput in my roguelike game to quite some success, but it is mostly for decoding UTF8 text input from the keyboard at this stage and not so much for widgets. That being said, it could still have some use to us in terms of the widget implementation.

**Comment by Mikolaj** [comment]
I'd rather do this in a way that works 100% the same in SDL and in the browser using character-input that already works fine, but if players request Japanese names (and contribute good fonts that contain such characters) I'm open to implementing this per-frontend.

**Comment by Mikolaj** [comment]
Fine, let's do this per-frontend with full UTF8 support.

Regarding contributing fonts, people can customize their fonts even in their own installation of a game, but I may not be generous enough to maintain or yeats a feature I don't use if they are not generous even as much as to share their font config customization or their game content for the benefit of other players or game creators.
